File: duquad_matlab/FGM.m
function [zopt,fopt,niter] = FGM(varargin)
% *** Fast Gradient method ***
%
% Solves the problem: f(x) = 0.5*x'*H*x + c'x,
% where x = [x1 x2 ... xN]^T
%
% INPUTS:
% H: input matrix (must be positive definite and symmetric)
% c: input vector
% lb: lower bound vector
% ub: upper bound vector
% z0: initial guess
% opt: struct containing algorithm options
% 
% OUTPUTS:
% xopt: optimal solution
% fopt: optimal value (f(xopt))
% niter: number of iteration for finding xopt
% 
% USAGE:
% [xopt,fopt,niter] = FG(H,c);
% [xopt,fopt,niter] = FG(H,c,lb,ub);
% [xopt,fopt,niter] = FG(H,c,lb,ub,z0);
% [xopt,fopt,niter] = FG(H,c,lb,ub,z0,opt);
% [xopt,fopt,niter] = FG(H,c,[],[],[],opt); 
% 
% AUTHOR:
% Sverre Kvamme

% TODO:
% - check if matrix dim er korrekte
% - sholud express if one of the stoppingcriterias are met
% - check for typos
% - make help text perfect
% - take the time
% - update the gradient_descent with the same changes

if nargin < 7
    info = [];
    if nargin < 6
        opt = [];
        if nargin < 5
            z0 = [];
            if nargin < 4
                ub = [];
                if nargin < 3
                    lb = [];
                    if nargin < 2
                        error('Incorrect number of input arguments.')                 
                    end
                end
            end
        end
    end
end
    
[H,c] = deal(varargin{1:2});

if nargin > 2
    lb = varargin{3};
end    
if nargin > 3
    ub = varargin{4};
end    
if nargin > 4
    z0 = varargin{5};
end  
if nargin > 5
    opt = varargin{6};
end
if nargin > 6
    info = varargin{7};
end
if nargin > 7
    error('Incorrect number of input arguments.')
end    

n = size(H,2);

if isempty(lb)
    lb_inf = true;
else
    lb_inf = false;
end
if isempty(ub)
    ub_inf = true;
else
    ub_inf = false;
end
if isempty(z0)
    z0 = zeros(n,1);
end
if isempty(opt)
    opt = set_options('default');
end
if isempty(info)
    H_eig = eig(H);
    info.L =  max(H_eig);
    info.mu = min(H_eig);
end
    
q = info.mu / info.L;
one_over_L = 1/info.L;
alpha = 0.5;

% initialize other necessary variables and vectors
diff_function_value = opt.eps+1; 
f_value = obj(z0,H,c);
z = z0;
y = z;

niter = 0;


while diff_function_value > opt.eps
   
    if niter > opt.maxiter
        %warning('reached maximum number of iterations in fast_gradient'
        break;
    end
    
    % calculating next z = y - (1/L)(H*y+c)
    znew = y - one_over_L * (H*y + c);%grad(y,H,c);
     
    % projection xnew on the feasible set made of lb and ub,
    % but do not project if problem is unconstrained 
    if ~ub_inf
        znew = min(ub,znew);
    end
    if ~lb_inf
        znew = max(lb,znew);
    end

    % calculating next y 
    alpha_pow2 = alpha*alpha;
    theta = q - alpha_pow2;
    alpha_new = 0.5 * (theta + sqrt(theta^2 + 4*alpha_pow2));
    beta = (alpha*(1-alpha)) / (alpha_pow2 + alpha_new);
    y = znew + beta*(znew - z);
        
    % calculate the difference between the new and previous function value
    f_value_new = obj(znew,H,c);
    diff_function_value = abs(f_value_new - f_value);

    % update the variables
    z = znew;
    f_value = f_value_new;
    alpha = alpha_new;
    niter = niter+1;
   
end
%fprintf('niter inner: %d\n', niter)
zopt = z;
fopt = f_value;

end

function f = obj(z, H, c)
f = 0.5 * (z'*H*z) + c'*z;
end

File: duquad_c/src/fgm.c
/*
 * FGM.c
 *
 *  Created on: Aug 21, 2014
 *      Author: Sverre
 */

#include "fgm.h"

// NOTE: z0 must be feasible

uint32_t FGM(struct Struct_FGM *s)
{

	s->exitflag = 1;

	// z = z0, y = z
	vector_copy(s->z0,s->z,N);
	vector_copy(s->z,s->y,N);

	// Initialize the function value, and difference between the new and previous function value
	real_t f = obj(s->z, s->H, s->c, s->temp1_dim_N);
	real_t fnew = 0.0;
	real_t f_diff = s->eps + 1;

	// Declaration and initialization of other necessary variables
	real_t L = s->eigH_max;
	real_t mu = s->eigH_min;
	real_t one_over_L = 1.0/L;
	real_t q = mu/L;
	real_t alpha = 0.5;
	real_t alpha_new = 0.0;
	real_t theta = 0.0;
	real_t beta = 0.0;
	real_t alpha_pow2 = 0.0;

	// iteration counter
	uint32_t niter = 0;


	/* ##################################################################################
			START WHILE-LOOP
	################################################################################## */
	while(f_diff > s->eps)
	{

		if (niter > s->maxiter){
			//fprintf(stderr, "reached maximum number of iterations in FGM (inner problem)\n");
			s->exitflag = 2;
			break;
		}

		/* *********************************************************************
	 			Finding znew
		********************************************************************* */

		// Calculating gradient f'(y) = H*y + c
		mtx_vec_mul(s->H,s->y,s->temp1_dim_N,N,N); // H * y
		vector_add(s->c,s->temp1_dim_N,s->temp1_dim_N,N); // + c

		// z_new = y - (1/L * f'(y))
		vector_scalar_mul(s->temp1_dim_N, one_over_L, s->temp1_dim_N, N); // 1/L * f'(y)
		vector_sub(s->y, s->temp1_dim_N, s->znew, N);	// y - ans

		// Projection znew on the feasible set made of lb and ub
		if(!s->ub_is_inf)
			vector_min(s->ub,s->znew,s->znew,N);
		if(!s->lb_is_inf)
			vector_max(s->lb,s->znew,s->znew,N);

		/* *********************************************************************
	 			Finding ynew
		********************************************************************* */

		// Calculating beta: beta = alpha(1-alpha) / alpha²+alpha_new,
		// where: alpha² = (1-alpha_new)*alpha² + q*alpha_new
		alpha_pow2 = alpha*alpha;
		theta = q - alpha_pow2;
		alpha_new = 0.5 * (theta + sqrt((theta*theta) + 4.0*alpha_pow2));
		beta = (alpha*(1.0-alpha)) / (alpha_pow2 + alpha_new);

		// ynew = znew + beta*(znew-z)
		vector_sub(s->znew,s->z,s->temp1_dim_N,N); // znew-z
		vector_scalar_mul(s->temp1_dim_N, beta, s->temp1_dim_N, N); // beta * (znew-z)
		vector_add(s->znew,s->temp1_dim_N,s->ynew,N); // + znew

		/* *********************************************************************
	 			Finding f_diff
		********************************************************************* */

		// Calculate the difference between the new and previous function value
		fnew = obj(s->znew, s->H, s->c, s->temp1_dim_N);
		f_diff = abs_2(fnew - f);

		/* *********************************************************************
	 			Update the variables
		********************************************************************* */

		vector_copy(s->znew,s->z,N);
		vector_copy(s->ynew,s->y,N);
		f = fnew;
		alpha = alpha_new;
		niter++;

	}	/* END WHILE-LOOP */

	vector_copy(s->znew,s->zopt,N);
	s->fopt = fnew;

	return niter;
}

// This function is not used by fgm or dfgm
void clean_up_FGM_C(struct Struct_FGM *s)
{
	free_pointer(s->H);
	free_pointer(s->c);
	free_pointer(s->lb);
	free_pointer(s->ub);
	free_pointer(s->z0);
	free_pointer(s->zopt);
	free_pointer(s->z);
	free_pointer(s->y);
	free_pointer(s->znew);
	free_pointer(s->ynew);
	free_pointer(s->temp1_dim_N);
}
